core/drivers/crypto/versal/hash.c

```diff
@@ -0,0 +1,246 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2022
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <assert.h>
+#include <drvcrypt.h>
+#include <drvcrypt_hash.h>
+#include <initcall.h>
+#include <ipi.h>
+#include <kernel/panic.h>
+#include <kernel/refcount.h>
+#include <kernel/spinlock.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+
+/* single engine */
+struct mutex lock = MUTEX_INITIALIZER;
+
+#define FIRST_PACKET	BIT(30)
+#define NEXT_PACKET	BIT(31)
+
+enum versal_sha3_state {
+	SHA3_STNDBY = 0, SHA3_INIT, SHA3_RUN
+};
+
+struct update_request {
+	uint8_t *data;
+	size_t len;
+	STAILQ_ENTRY(update_request) link;
+};
+
+struct versal_hash_ctx {
+	struct crypto_hash_ctx hash_ctx;
+	enum versal_sha3_state state;
+	STAILQ_HEAD(update_request_list, update_request) req_list;
+};
+
+static const struct crypto_hash_ops versal_ops;
+
+static struct versal_hash_ctx *to_versal_ctx(struct crypto_hash_ctx *ctx)
+{
+	assert(ctx && ctx->ops == &versal_ops);
+
+	return container_of(ctx, struct versal_hash_ctx, hash_ctx);
+}
+
+static TEE_Result do_hash_init(struct crypto_hash_ctx *ctx)
+{
+	struct versal_hash_ctx *c = to_versal_ctx(ctx);
+
+	if (c->state != SHA3_STNDBY)
+		return TEE_ERROR_GENERIC;
+
+	c->state = SHA3_INIT;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result do_hash_update(struct crypto_hash_ctx *ctx,
+				 const uint8_t *data, size_t len)
+{
+	struct versal_hash_ctx *c = to_versal_ctx(ctx);
+	struct update_request *req = NULL;
+
+	/*
+	 * The engine does not have the concept of context, so all update
```

Comment:
It seems that this implementation should target a special use case and should normally be avoided in a generic case where we don't have control over how much data is to be hashed.

If this is accessible from user space it can't be too hard to cause a panic somewhere in this driver.

Reply:
Right, I did my best to provide standard functionality with the limitations it has. WRT panic, I suppose as easy as making OP-TEE panic...not sure if that is a bad thing as long as it is a controlled panic? sorry not sure what do you mean, maybe I miss your point

Reply:
Given how limited this is perhaps it's time to take a step back and consider the use cases we'd like to cover with this driver. For general purpose usage, I guess the system will panic so often that it will be more or less impossible to work with.

Reply:
Not sure about that, in fact I doubt the standard user who allocates a context, processes and leaves will hit any issue.

https://github.com/ldts/optee_os/blob/xlnx_v2022.1.versal.001/core/drivers/crypto/versal/hash_test.c this is what I did to test.

Reply:
The way to make it pass is to pass a tiny payload. Try to check that hash of a 2meg TA or something.

Reply:
yes that wouldn't be good. That would have been a good test to have in xtest. But yes, running out of memory is a risk with this architecture.

Having said that how is that use case is  different from say creating keys using the the pkcs#11 TA?. It is equally easy to panic the TA by creating keys in a loop. Is this a bug then?

But I understand your point. Unfortunately I have no work-around that...so what do you propose? shall I just drop it?

Reply:
A TA that panics isn't good, but if we panic and hang the entire system that's not good on a different level.

> But I understand your point. Unfortunately I have no work-around for that...so what do you propose? shall I just drop it?

Perhaps that's best for now since it can't be used as the primary generic hash implementation.

Reply:
ok.

SHA3-384 was being used by the Versal HUK driver (which I'll post after these) but I can do it a bit differently

Reply:
Makes sense

Reply:
> yes that wouldn't be good. That would have been a good test to have in xtest. But yes, running out of memory is a risk with this architecture.
> 
> Having said that, how is that use case different from say creating keys using the the pkcs#11 TA?. It is equally easy to panic the TA by creating keys in a loop. Is this a bug then?
> 
> But I understand your point. Unfortunately I have no work-around for that...so what do you propose? shall I just drop it?

Hi @ldts specifically where are you concerned about running out of memory and which memory in particular? There's quite a bit of memory to chose from in Versal.

Reply:
hi @nathan-menhorn. right so since the hardware engine doesn't keep the context, the TEE needs to keep contexts requests and updates in its internal memory until the active context finalizes. 
Imagine a thread starts an operation, and then stops while other threads start queuing sha3-384 requests. We will end up starving the memory available in the TEE.

Having said that, maybe we could just associate a memory pool to the sha3 driver if you think this driver is important - AFAIR it is main use case is for the HUK and not much more so I was planing of moving it to core/drivers so it is just internal to the TEE.

@jenswi-linaro would you be ok if I allocated a memory pool for this driver?

Reply:
the idea is that big payloads will not affect the rest of the system, just the operation

Reply:
> @jenswi-linaro would you be ok if I allocated a memory pool for this driver?

I don't see how you can make that work. Whatever amount of memory you sacrifice you will still be able to run out of it.

Reply:
> hi @nathan-menhorn. right so since the hardware engine doesn't keep the context, the TEE needs to keep contexts requests and updates in its internal memory until the active context finalizes. Imagine a thread starts an operation, and then stops while other threads start queuing sha3-384 requests. We will end up starving the memory available in the TEE.
> 
> Having said that, maybe we could just associate a memory pool to the sha3 driver if you think this driver is important - AFAIR it is main use case is for the HUK and not much more so I was planing of moving it to core/drivers so it is just internal to the TEE.
> 

Hi @ldts that makes sense. There's use cases outside of the HUK but moving it core/drivers makes sense. However, I also see you are working on a shared memory pool. I'll just keep an eye on this PR. Thanks!

Reply:
@jenswi-linaro I added some info to the commit messagge  https://github.com/OP-TEE/optee_os/pull/5643/commits/46f13a439273beea37071098d5f663b069484b99

Reply:
it seems acceptable to me.

Reply:
No, this can't be used as a generic hash to the GlobalPlatform TEE Internal Core API.
It seems that this hardware can only be used in a more limited use case.

Reply:
yeah I now fully agree. I tried to implement support for multithread and is just a huge pain on top of the memory consumption...it just wont scale. ok will move to core/drivers as a companion driver